File: app/src/main/java/com/slimshady/bottomsheetmodule/adapter/BottomSheetRecyclerAdapter.java
package com.slimshady.bottomsheetmodule.adapter;


import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.slimshady.bottomsheetmodule.adapter.listener.AdapterItemListener;
import com.slimshady.bottomsheetmodule.databinding.ItemBaseSearchBinding;
import com.slimshady.bottomsheetmodule.model.BaseBottomSheetRecyclerModel;
import com.slimshady.bottomsheetmodule.util.AdapterAction;

import java.util.ArrayList;

public class BottomSheetRecyclerAdapter<T extends BaseBottomSheetRecyclerModel> extends RecyclerView.Adapter<BottomSheetRecyclerAdapter.ViewHolder> {

    private ArrayList<T> items;
    private boolean isMultiSelect;
    private AdapterItemListener<T> adapterItemListener;

    public BottomSheetRecyclerAdapter(ArrayList<T> items, boolean isMultiSelect, AdapterItemListener<T> adapterItemListener) {
        this.items = items;
        this.isMultiSelect = isMultiSelect;
        this.adapterItemListener = adapterItemListener;
    }

    public BottomSheetRecyclerAdapter(ArrayList<T> items, boolean isMultiSelect) {
        this.items = items;
        this.isMultiSelect = isMultiSelect;
    }

    @NonNull
    @Override
    public BottomSheetRecyclerAdapter<T>.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemBaseSearchBinding itemBaseSearchBinding =
                ItemBaseSearchBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new BottomSheetRecyclerAdapter<T>.ViewHolder(itemBaseSearchBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull BottomSheetRecyclerAdapter.ViewHolder holder, int position) {
        holder.bind(items.get(position));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ItemBaseSearchBinding binding;

        public ViewHolder(@NonNull ItemBaseSearchBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(T baseSearchDbModel) {
            binding.txtSearch.setText(baseSearchDbModel.getName());

            if (isMultiSelect)
                binding.checkboxSearch.setVisibility(View.VISIBLE);
            else
                binding.checkboxSearch.setVisibility(View.GONE);



            binding.getRoot().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (isMultiSelect)
                        toggleSelection(baseSearchDbModel, binding);
                    else
                        adapterItemListener.onItemSelect(baseSearchDbModel, getAbsoluteAdapterPosition(), AdapterAction.SELECT);
                }
            });


        }
    }

    boolean isSelected(int position) {
        return items.get(position).isSelected();
    }

    public void toggleSelection(T baseSearchDbModel, ItemBaseSearchBinding binding) {
        if (!baseSearchDbModel.isSelected()) {
            baseSearchDbModel.setSelected(true);
            binding.checkboxSearch.setChecked(true);

        } else {
            baseSearchDbModel.setSelected(false);
            binding.checkboxSearch.setChecked(false);

        }

    }

    public void selectAll() {
        for (int i = 0; i < items.size(); i++) {
            items.get(i).setSelected(true);

        }
    }

    public void clearSelections() {
        for (int i = 0; i < items.size(); i++) {
            items.get(i).setSelected(false);

        }

    }

    public int getSelectedItemCount() {
        int count = 0;

        for (int i = 0; i < items.size(); i++) {
            if (items.get(i).isSelected())
                count++;

        }

        return count;
    }

    public ArrayList<T> getSelectedItems() {
        ArrayList<T> mDifferTemp = new ArrayList<>();

        for (int i = 0; i < items.size(); i++) {
            if (items.get(i).isSelected())
                mDifferTemp.add(items.get(i));

        }
        return mDifferTemp;

    }
}

File: app/src/main/java/com/slimshady/bottomsheetmodule/adapter/BottomSheetRecyclerAdapter.kt
package com.slimshady.bottomsheetmodule.adapter

import com.slimshady.bottomsheetmodule.model.BaseBottomSheetRecyclerModel
import androidx.recyclerview.widget.RecyclerView
import com.slimshady.bottomsheetmodule.adapter.listener.AdapterItemListener
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import com.slimshady.bottomsheetmodule.databinding.ItemBaseSearchBinding
import com.slimshady.bottomsheetmodule.util.AdapterAction
import java.util.ArrayList

class BottomSheetRecyclerAdapter<T : BaseBottomSheetRecyclerModel?> :
    RecyclerView.Adapter<BottomSheetRecyclerAdapter<T>.ViewHolder> {
    private var items: ArrayList<T>
    private var isMultiSelect: Boolean
    private var adapterItemListener: AdapterItemListener<T>? = null

    constructor(
        items: ArrayList<T>,
        isMultiSelect: Boolean,
        adapterItemListener: AdapterItemListener<T>?
    ) {
        this.items = items
        this.isMultiSelect = isMultiSelect
        this.adapterItemListener = adapterItemListener
    }

    constructor(items: ArrayList<T>, isMultiSelect: Boolean) {
        this.items = items
        this.isMultiSelect = isMultiSelect
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemBaseSearchBinding =
            ItemBaseSearchBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(itemBaseSearchBinding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(items[position])
    }

    override fun getItemCount(): Int {
        return items.size
    }

    inner class ViewHolder(var binding: ItemBaseSearchBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        fun bind(baseSearchDbModel: T) {
            binding.txtSearch.text = baseSearchDbModel?.name
            if (isMultiSelect) binding.checkboxSearch.visibility =
                View.VISIBLE else binding.checkboxSearch.visibility = View.GONE
            binding.root.setOnClickListener {
                if (isMultiSelect) toggleSelection(
                    baseSearchDbModel,
                    binding
                ) else adapterItemListener?.onItemSelect(
                    baseSearchDbModel,
                    absoluteAdapterPosition,
                    AdapterAction.SELECT
                )
            }
        }
    }

    fun isSelected(position: Int): Boolean {
        return items[position]?.isSelected ?: false
    }

    fun toggleSelection(baseSearchDbModel: T, binding: ItemBaseSearchBinding) {
        if (baseSearchDbModel?.isSelected == true) {
            baseSearchDbModel.isSelected = false
            binding.checkboxSearch.isChecked = false

        } else {
            baseSearchDbModel?.isSelected = true
            binding.checkboxSearch.isChecked = true

        }
    }

    fun selectAll() {
        for (i in items.indices) {
            items[i]?.isSelected = true
        }
    }

    fun clearSelections() {
        for (i in items.indices) {
            items[i]?.isSelected = false
        }
    }

    val selectedItemCount: Int
        get() {
            var count = 0
            for (i in items.indices) {
                if (items[i]?.isSelected == true) count++
            }
            return count
        }
    val selectedItems: ArrayList<T>
        get() {
            val mDifferTemp = ArrayList<T>()
            for (i in items.indices) {
                if (items[i]?.isSelected == true) mDifferTemp.add(items[i])
            }
            return mDifferTemp
        }
}